### PR TITLE
Add wild card support for $InterfaceAlias.

### PR DESCRIPTION
**Pull request by guitarrapc**
Description
----

This is re-create PR from https://github.com/PowerShell/xNetworking/pull/56 to be merged to dev branch.

There used to be support Wildcard for InterfaceAlis, but degraded with commit https://github.com/PowerShell/xNetworking/commit/89a9d95b611782a1848f7c9a3a66d550ab0a4afa#diff-da1b602b4060c5ac9213c9b402ca75aaR178

This PR add Wildcard support for InterfaceAlias.

There are many cases that user couldn't know or promise interface alias. Something like it is  "Ethernet" or "Ethernet 2" or "Enternet n..." in Cloud Infrastructure. Thus Wildcard support is useful and should be better supported.

Reproduce Procedure
----

```powershell
configuration Hoge
{
    Import-DSCResource -Module xNetworking
    xDNSServerAddress Test
    {
        Address = "10.0.0.10"
        InterfaceAlias = "Etthernet*"
        AddressFamily = "IPv4"
    }
}

Hoge
Start-DscConfiguration -Verbose -Wait -Path Hoge -Force
Test-DscConfiguration
Get-DscConfiguration
```

It will fail with previous psm1, but success in this PR as following.

```
Address        : {10.0.0.10}
AddressFamily  : IPv4
InterfaceAlias : Ehternet*
PSComputerName : 
```

### CODE REVIEW

**Comment by guitarrapc** [comment]
Sorry for late response, I will do it right now.

- [ ] Move Comment
- [ ] Write Unit Test : Unit\xDNSServerAddress
- [ ] Adapt DNSServer change to multiple Adaptor with foreach loop.

**Comment by PlagueHO** [comment]
Awesome! :+1: Thank you!

**Comment by guitarrapc** [comment]
I figure out why Unity Test was failed with https://github.com/guitarrapc/xNetworking/commit/fb9ffac876cba5d0d25c3a27789c5bb2a648fbb6.

This is because ```Get-NetAdapter``` sat Mock with ```Name = 'Ethernet'```
https://github.com/PowerShell/xNetworking/blob/d43b2f6bda9a7bbb63bd1ff7db40191a22e8c64d/Tests/Unit/MSFT_xDNSServerAddress.Tests.ps1#L328

As of Cmdlet output will be limit to what MockWith has passed, InterfaceAlias detected null in UnitTest. This is no matter on real code, but only issue in MockWith.

I think there are two workaround.

1. Change InterfaceAlias to Name Property for ```Get-NetAdapter``` result.
  Because it seems always Name == InterfaceAlias, I choose this.
2. Comment out UnitTest for ```Get-NetAdapter``` using MockWith https://github.com/PowerShell/xNetworking/blob/d43b2f6bda9a7bbb63bd1ff7db40191a22e8c64d/Tests/Unit/MSFT_xDNSServerAddress.Tests.ps1#L328

Which do you prefer, or is there any good idea for it?

**Comment by guitarrapc** [comment]
All test passed:) I think it works fine and already working in my production.

Please have a kind review, thanks.

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 1]
```diff
@@ -161,34 +165,41 @@ function Test-TargetResource
...
+                -DifferenceObject $Address `
+                -SyncWindow 0).Length -gt 0)
+
+        if ($addressDifferent)
+        {
+            $desiredConfigurationMatch.Add($false)
+            Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
+                $($LocalizedData.DNSServersNotCorrectMessage) `
+                    -f ($Address -join ','),($currentAddress -join ',')
+                ) -join '' )
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: One thought woudl be to just return $false here and exit out of the loop, as we're already going to need to make changes not sure it's really worth the time to check everything.

**Comment by tysonjhayes** [comment]
Thanks for the PR, I'll be reviewing this a little later, glancing through it it looks pretty good.

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 1 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Yes it is one of way. But I though it worth Write-Verbose will tell which interface was not done and which is done.

It depend on which do you like "Fail first" or "Run all and detect all at once". I will follow you.

Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Good point that I hadn't considered. Let's stick with the current model.

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 2]
```diff
@@ -48,15 +48,14 @@ function Get-TargetResource
         $($LocalizedData.GettingDNSServerAddressesMessage)
         ) -join '')
 
+    $dnsClientServerAddresses = Get-NetAdapter -Name $InterfaceAlias | ForEach-Object { Get-DnsClientServerAddress -InterfaceAlias $_  -AddressFamily $AddressFamily }
```
Comment (PlagueHO) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Any chance we could get this line down below the 100 char line limit?

e.g.
```powershell
$dnsClientServerAddresses = Get-NetAdapter -Name $InterfaceAlias `
 | ForEach-Object { Get-DnsClientServerAddress -InterfaceAlias $_  -AddressFamily $AddressFamily }
```

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 3]
```diff
@@ -161,34 +165,41 @@ function Test-TargetResource
...
+                $($LocalizedData.DNSServersSetCorrectlyMessage)
+                ) -join '' )
+        }
     }
-    return $desiredConfigurationMatch
+
+    return !($desiredConfigurationMatch -contains $false)
 }
 
 #######################################################################################
```
Comment (PlagueHO) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: While you're making changes, would you mind removing these ############################## separator lines?

I added them, but they are not useful and we're trying to standardize the code style :smile:

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 4]
```diff
@@ -48,15 +48,14 @@ function Get-TargetResource
...
+        Address = $dnsClientServerAddresses.ServerAddresses
         AddressFamily = $AddressFamily
         InterfaceAlias = $InterfaceAlias
     }
 
-    $returnValue
+    return $returnValue
 }
 
 ######################################################################################
```
Comment (PlagueHO) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: While you're making changes, would you mind removing these ############################## separator lines?

I added them, but they are not useful and we're trying to standardize the code style :smile:

**Comment by PlagueHO** [comment]
The new code looks great! :+1: I can't see any issues.

One request: is it possible to add some tests where the Get-NetAdapter mock returns more than one record (as would occur when a wild card was specified and more than one adapter exists)? For example:
```powershell
Mock Get-NetAdapter -MockWith { @(
[PSObject]@{ Name = 'Ethernet 1' }
[PSObject]@{ Name = 'Ethernet 2' }
) }
```

I don't think this isn't critical though. @tysonjhayes - what are your thoughts? Worth adding these tests?

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 2 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Yes, I can change it. But why are there 100 char line limitation? It's just my curious, but any performance issue with line char length? Or just a good looking?

**Comment by guitarrapc** [comment]
@PlagueHO You are right! I also think it should read for multiple interface card with Mock. I will do it.

**Comment by guitarrapc** [comment]
Todo
----

- [ ] 100 char line limitation :  https://github.com/PowerShell/xNetworking/pull/57#discussion_r46199922
- [ ] Standardize the code style for regardless # line.
- [ ] Multiple Interface Card Mocking Unit Test : https://github.com/PowerShell/xNetworking/pull/57#issuecomment-160759384

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 2 above)
Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: [Style guidance](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md) of our resources. So we're just adhearing to it, it's to make code reviews easier on smaller monitors I think.

Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Appreciate. Almost all style guidance was what I'm always doing, without 100 char line. I will follow it.

Comment (PlagueHO) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: :+1: Thank you!

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 4 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: done with commit https://github.com/guitarrapc/xNetworking/commit/3e2d04de231228b55c4ce6171033fdb8ceb57938

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 3 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: done with commit https://github.com/guitarrapc/xNetworking/commit/3e2d04de231228b55c4ce6171033fdb8ceb57938

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 2 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: done with commit https://github.com/guitarrapc/xNetworking/commit/3e2d04de231228b55c4ce6171033fdb8ceb57938

Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: done https://github.com/guitarrapc/xNetworking/commit/510b89347338f500c1630187125ec9cca5fdabec

**Comment by guitarrapc** [comment]
Alright, all todo had been completed without test failure.

**Comment by PlagueHO** [comment]
Awesome! Thank you so much for working on this. :+1:

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 5]
```diff
@@ -87,73 +105,86 @@ function Set-TargetResource
...
-            InterfaceAlias = $InterfaceAlias
-            Address = $Address
-            Validate = $Validate
+        $currentAddress = (Get-DnsClientServerAddress `
+                -InterfaceAlias $alias `
+                -AddressFamily $AddressFamily `
+                -ErrorAction Stop).ServerAddresses
+
+        # Check if the Server addresses are the same as the desired addresses.
+        [Boolean] $addressDifferent = (@(Compare-Object `
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Thinking through this again, do we really need to do this? If Test was true then we wouldn't be running set. As we wouldn't be running set do we need this if/else logic? If not could we just set the DNS Client and be done with it?

**Comment by tysonjhayes** [comment]
Ok for some reason the tests aren't showing the diff which is making it hard to review. @PlagueHO I know you've run into that at some point, what did you do to fix it?

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 (hunk 5 above)
Comment (guitarrapc) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: I think you mean to change Set-TargetResource like this. (remove if clause.)

```powershell
function Set-TargetResource
{
<#
.SYNOPSIS 
Set Node configuration to desired state.

.DESCRIPTION
Set a new Server Address in the current node
#>
    [CmdletBinding()]
    param
    (    
        # IP Address that has to be set    
        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [String[]]$Address,

        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [String]$InterfaceAlias,

        [Parameter(Mandatory = $true)]
        [ValidateSet('IPv4', 'IPv6')]
        [String]$AddressFamily,
        
        [Boolean]$Validate = $false
    )

    Write-Verbose -Message ( @("$($MyInvocation.MyCommand): "
        $($LocalizedData.ApplyingDNSServerAddressesMessage)
        ) -join '')

    # Get the current DNS Server Addresses based on the parameters given.
    # Interface Alias may return multiple when passed WildCard (*)
    $interfaceAliases = Get-NetAdapter -Name $InterfaceAlias
    foreach ($alias in $interfaceAliases.InterfaceAlias)
    {
        # Set the DNS settings as well
        $Splat = @{
            InterfaceAlias = $alias
            Address = $Address
            Validate = $Validate
        }
        Set-DnsClientServerAddress @Splat `
            -ErrorAction Stop
    }
}
```

Agree. But how we handle which interface don't need to set? Do you accept change both "already configured DNS" and "not configured DNS" when 1 of interface required to be sat?
If so, it should be better fail fast on Test-TargetResource too,
https://github.com/guitarrapc/xNetworking/blob/AddWildcardSupport/DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1#L211-L227

Something like this.

```powershell
function Test-TargetResource
{
<#
.SYNOPSIS 
Test Node configuration is in desired state or not.
.DESCRIPTION
Test if the given Server Address is among the current node's Server Address collection.
#>
    [OutputType([System.Boolean])]
    [CmdletBinding()]
    param
    (        
        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [String[]]$Address,

        [Parameter(Mandatory = $true)]
        [ValidateNotNullOrEmpty()]
        [String]$InterfaceAlias,

        [Parameter(Mandatory = $true)]
        [ValidateSet('IPv4', 'IPv6')]
        [String]$AddressFamily,
        
        [Boolean]$Validate = $false        
    )

    Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
        $($LocalizedData.CheckingDNSServerAddressesMessage)
        ) -join '' )

    # Validate the Settings passed
    Foreach ($ServerAddress in $Address) {       
        Test-ResourceProperty `
            -Address $ServerAddress `
            -AddressFamily $AddressFamily `
            -InterfaceAlias $InterfaceAlias
    }

    # Get the current DNS Server Addresses based on the parameters given.
    # Interface Alias may return multiple when passed WildCard (*)
    $interfaceAliases = Get-NetAdapter -Name $InterfaceAlias
    foreach ($alias in $interfaceAliases.Name)
    {
        $currentAddress = (Get-DnsClientServerAddress `
            -InterfaceAlias $alias `
            -AddressFamily $AddressFamily `
            -ErrorAction Stop).ServerAddresses

        # Check if the Server addresses are the same as the desired addresses.
        [Boolean] $addressDifferent = (@(Compare-Object `
                -ReferenceObject $currentAddress `
                -DifferenceObject $Address `
                -SyncWindow 0).Length -gt 0)

        if ($addressDifferent)
        {
            # fail fast return when one of interface Test result was false.
            $desiredConfigurationMatch.Add($false)
            Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
                $($LocalizedData.DNSServersNotCorrectMessage) `
                    -f ($Address -join ','),($currentAddress -join ',')
                ) -join '' )
            return $false;
        }
        else 
        {
            # Test will return true in this case
            Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
                $($LocalizedData.DNSServersSetCorrectlyMessage)
                ) -join '' )
        }
    }
}
```

**Comment by PlagueHO** [comment]
Ah, right. From memory (I couldn't find the PR) I actually closed the PR and actually remade the changes to the tests file and then reissued the PR. That might be difficult here though as my changes were fairly trivial compared to these. There must be a reason for this odd behavior from GitHub. I'll take a dig around today and see if I can find a solution that doesn't involve closing the PR.

**Comment by guitarrapc** [comment]
May be this? http://stackoverflow.com/questions/9702795/large-github-commit-diff-not-shown

**Comment by PlagueHO** [comment]
It does look like an issue (or limitation) with GitHub. Although I'm fairly certain your Diff is well within the 100KB limit. Might have to revert to doing a manual Git Diff directly from your repo.

**Comment by tysonjhayes** [comment]
Sorry about the delay here, I've pulled it locally and will be looking a it as I have time. Unfrotunately I've been a bit busy today and haven't had time yet.

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 6]
```diff
@@ -47,36 +55,46 @@ function Get-TargetResource
     Write-Verbose -Message ( @( "$($MyInvocation.MyCommand): "
         $($LocalizedData.GettingDNSServerAddressesMessage)
         ) -join '')
+    
+    # obtain all the interface addresses
+    $dnsClientServerAddresses = Get-NetAdapter -Name $InterfaceAlias `
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Could you help me understand this change better? I'm not sure I'm clear on why its needed? Also when testing it on my machine I get an error when running this command.

````powershell
[4:29:27 PM] [pr-57] [0057] » Get-TargetResource -Address '127.0.0.1' -InterfaceAlias '*Ethernet*' -AddressFamily 'IPv4'
Get-DnsClientServerAddress : No matching MSFT_DNSClientServerAddress objects found by CIM query for instances of the ROOT/StandardCimv2/MSFT_DNSClientServerAddress class on the  CIM server: SELECT * FROM MSFT_DNSClientServerAddress
WHERE ((InterfaceAlias LIKE 'MSFT[_]NetAdapter (CreationClassName = "MSFT[_]NetAdapter", DeviceID = "{EB44F4BF-30C1-401E-821A-AAAF8229A5E6}", SystemCreationClassName = "CIM[_]NetworkPort", SystemName =
"<ComputerName>)')) AND ((AddressFamily = 2)). Verify query parameters and retry.
At C:\GitHub\tysonjhayes\xNetworking\DSCResources\MSFT_xDNSServerAddress\MSFT_xDNSServerAddress.psm1:61 char:24
+     | ForEach-Object { Get-DnsClientServerAddress -InterfaceAlias $_  -AddressFa ...
+                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (MSFT_DNSClientServerAddress:String) [Get-DnsClientServerAddress], CimJobException
    + FullyQualifiedErrorId : CmdletizationQuery_NotFound,Get-DnsClientServerAddress
````

I get the same thing when I don't try a wild card for the interface alias. Where the previous comamnd worked fine.

File: DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1 [hunk 7]
```diff
@@ -47,36 +55,46 @@ function Get-TargetResource
...
-# The Set-TargetResource cmdlet.
-# This function will set a new Server Address in the current node
-######################################################################################
+#endregion
+
+#region Set-TargetResource function
+
 function Set-TargetResource
 {
+<#
```
Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Could we modify this to:
````powershell
function Test-TargetResource
{
    <#
        .SYNOPSIS 
        Test Node configuration is in desired state or not.

        .DESCRIPTION
        Test if the given Server Address is among the current node's Server Address collection.
    #>
````

That way it collapses nicely in most code folding editors.

Comment (tysonjhayes) [inline DSCResources/MSFT_xDNSServerAddress/MSFT_xDNSServerAddress.psm1]: Also do the same for the other functions.

**Comment by tysonjhayes** [comment]
OK in regards to the unit tests.

Couple of notes.

1) Let's remove the ####### lines and replace them with regions.
2) Can we change all of the "Describe 'Function' {" to "Describe 'MSFT_xDNSServerAddress\Function' {" so we know which one is failing in our stream of tests. h/t @PlagueHO 
3) Given I actually had a problem running the code on my machine but the unit tests are passing would it make since to throw a redumentary integration test along with this so we can validate everything is working as expected?

**Comment by tysonjhayes** [comment]
Unit Tests - as I can't comment on a diff.

## Get-TargetResource

````powershell

        Context 'invoking with an IPv4 address' {
            It 'should return true' {

                $Splat = @{
                    Address = '192.168.0.1'
                    InterfaceAlias = 'Ethernet'
                    AddressFamily = 'IPv4'
                }
                $Result = Get-TargetResource @Splat
                $Result.IPAddress | Should Be $Splat.IPAddress
            }
        }

````

This test is passing because IPAddress is null in both cases, I believe this should be:
````powershell
        Context 'invoking with an IPv4 address' {
            It 'should return true' {

                $Splat = @{
                    Address = '192.168.0.1'
                    InterfaceAlias = 'Ethernet'
                    AddressFamily = 'IPv4'
                }
                $Result = Get-TargetResource @Splat
                $Result.Address | Should Be $Splat.Address
            }
        }
````

The same goes for 
````powershell
        Context 'invoking with Wildcard InterfaceAlias and an IPv4 address' {
            It 'should return true' {

                $Splat = @{
                    Address = '192.168.0.1'
                    InterfaceAlias = 'Ether*'
                    AddressFamily = 'IPv4'
                }
                $Result = Get-TargetResource @Splat
                $Result.IPAddress | Should Be $Splat.IPAddress
            }
        }
````

In fact you can pretty much do a find/replace for IPAddress and replace it with Address.

## Set-TargetResource

Around the following code (Line 346)
````powershell
Context 'invoking with Wildcard InterfaceAlias && single IPv4 Server Address that is the same as current'
````

You start calling Get-NetAdapter. From what I can tell in current scoping you haven't mocked Get-NetAdapter so you'd be calling the machines' Get-NetAdapter, however you do Mock it in other tests. Let's add the mock everywhere so we don't call the machines adapters.


## Test-TargetResource

Later on (line 1439-1457) we have cases like this:

````powershell
        Mock Get-NetAdapter -MockWith { [PSObject]@{ Name = 'Ether*' } }
        Mock Get-DnsClientServerAddress -MockWith {

            [PSCustomObject]@{
                ServerAddresses = @('fe80:ab04:30F5:002b::1')
                InterfaceAlias = 'Ether*'
                AddressFamily = 'IPv6'
            }
        }
        #endregion

        Context 'invoking with Wildcard InterfaceAlias && Multiple interface && single IPv6 Server Address that is the same as current' {
            # Mock should capsul inside script block scope
            Mock Get-NetAdapter {
                return @(
                    [PSCustomObject]@{ Name = 'Ethernet 1'; InterfaceAlias = 'Ethernet 1' },
                    [PSCustomObject]@{ Name = 'Ethernet 2'; InterfaceAlias = 'Ethernet 2' }
                )
            }
````

Correct me if I'm wrong here but I don't think we need to call the first Mock and honestly is kind of confusing anyway.

**Comment by TravisEz13** [comment]
Please resolve merge conflicts.  For more information see [GitHowTo - 30. Resolving Conflicts](http://githowto.com/resolving_conflicts )

**Comment by TravisEz13** [comment]
If there is no update of the pull request within 7 days, we will close the pull request.  After that, if you have an update, feel free to reopen the pull request or a new pull request.

**Comment by guitarrapc** [comment]
oh.... Sorry for no response. I will review.

**Comment by TravisEz13** [comment]
@guitarrapc No problem, 
We appreciate the contribution.
We just try to manage the number of PR's we are reviewing.  So we closing them after there are no code updates for 21 days.  I'll give you a little more time, but please update, or close this and re-open when you are ready for it to be reviewed.  

Thanks, Travis

**Comment by guitarrapc** [comment]
OK, I'm in busy for moment, let's close for a while and open again. Thanks.

**Pull request closed without merging**